uast/uast.go

```diff
@@ -262,87 +270,311 @@ type Any interface{}
 type Identifier struct {
 	GenNode
+	// Name of an entity. Can be any valid UTF8 string.
 	Name string `json:"Name"`
 }
 
+// Roles returns a list of UAST node roles that this node implies.
+func (Identifier) Roles() []role.Role {
+	return []role.Role{
+		role.Identifier,
+	}
+}
+
+// String is an unescaped UTF8 string literal.
+//
+// What is considered a String literal:
+// - escaped string literals;
+// - raw string literals;
+// - path-like or url-like package names;
+//
+// Not considered a String literal:
+// - identifiers (see Identifier);
+// - qualified names (see QualifiedIdentifier);
+// - numeric and boolean literals;
+// - special regexp literals;
 type String struct {
 	GenNode
-	Value  string `json:"Value"`
-	Format string `json:"Format"` // TODO: make an enum later
+	// Value is a UTF8 string literal value.
+	//
+	// Drivers should remove any quotes and unescape the value according to the language rules.
+	Value string `json:"Value"`
+
+	// Format is an optional language-specific string that describes the format of the literal.
+	//
+	// This field can be empty for the most common string literal type of a specific language.
+	// The priority is given to a one-line literal that escapes newline characters.
+	//
+	// TODO: define some well-known formats and maybe make it an enum
+	Format string `json:"Format"`
 }
 
+// QualifiedIdentifier is a name of an entity that consists of multiple simple identifiers,
+// organized in a hierarchy, similar to filesystem paths.
+//
+// What is considered a QualifiedIdentifier:
+// - qualified names that consist of Identifier-like elements;
+//
+// Not considered a QualifiedIdentifier:
+// - path-like or url-like package names (see String);
+// - selector expressions (a->b and a.b in C++);
 type QualifiedIdentifier struct {
 	GenNode
+	// Names is a list of simple identifiers starting from a root level of hierarchy
+	// and ending with leaf identifier.
 	Names []Identifier `json:"Names"`
 }
 
+// Comment is a no-op node that can span multiple lines and provides a human-readable
+// description for code around it.
+//
+// TODO: currently some annotations are also considered a Comment; need to clarify this
 type Comment struct {
 	GenNode
-	Text   string `json:"Text"`
+
+	// Block is set to true for block-style comments.
+	//
+	// TODO: should be a string similar to Format field in String literal;
+	//       may have more than 2 possible values (line, block, doc?)
+	Block bool `json:"Block"`
+
+	// Text is an unescaped UTF8 string with the comment text.
+	//
+	// Drivers should trim any comment-related tokens as well as whitespaces and
+	// stylistic characters at the beginning of each line. See Prefix, Suffix, Tab.
+	//
+	// Example:
+	//    /*
+	//     * some comment
+	//     */
+	//
+	//    only "some comment" is considered a text
+	Text string `json:"Text"`
+
+	// Prefix is usually a set of whitespaces that appear before the first line of an
```

Comment:
What if it's not consistent? E.g.,

```
/*
 * Some comment:
 *     - Hello
 *     - Goodbye
 * some more comment.
 */
```
The actual first line has `""`, the first interesting line has `" "`, and so on.

Reply:
Your example is still pretty consistent.

The first line has "\n" which qualifies as a Prefix. Or other lines won't include it.
Following lines all have " * " as a Tab. Still consistent.
And the last line has "\n " as a Suffix. Again, Tab does not apply here and Suffix does not affect other lines.

It seems like the naming is bad though. I guess the Suffix appears to be a suffix of each line instead of the whole comment. Is that what you meant by this concern?

Reply:
The second case won't have Tab set at all, since, as you mentioned, it's not consistent. And there is no minimal prefix for each line that matches all the lines.

Currently, we have a function in `uast/transform` that does the splitting for all the drivers that request it. Although it's not implemented yet to a full degree (Tab is not set).